### ISSUE TITLE
Chatbot Support

### ISSUE THREAD

**Issue by pojntfx** [nextcloud/spreed#3259]
**Is your feature request related to a problem? Please describe.**
Due to the current COVID-19 crisis here in Germany and elsewhere, many more users than before are using Nextcloud Talk - I administer an instance for my high school, and the usage has jumped from 2-4 users _per week_ to 60-70 *concurrent* users _per day_. No matter how big I make the cloud instance (a signaling server is installed), it is almost impossible to get this to work with Nextcloud Talk conference calls (30 people per class, 16 classses in total). To work around this, I've integrated Nextcloud Talk with another FLOSS project, Jitsi, which is easily able to handle the load. This is done by the use of a chatbot; to write the latter, I've also implemented a simple Nextcloud Talk chat bot framework in Go as well as before-mentioned chatbot: https://github.com/pojntfx/nextcloud-talk-jitsi-bot.

This is currently done by using a polling mechanism, however this is more than just inefficient; moreover, it is required to add the chat bot user to every conversation that should have Jitsi integration. And no, `commands` are not the answer - almost no one (at least not a single person that I know of) has SSH access to their Nextcloud instance, mostly because they run their instance in Docker. It's 2020, we shouldn't have to (or even be able to; cloud native stuff is supposed to ephermal and immutable, after all) access a terminal to use chatbots; after all, this is a feature that almost all public chat software (i.e. MS Teams) has.

**Describe the solution you'd like**
It would be awesome if there were a framework for creating chatbots. This will most likely depend on some kind of streaming implementation, i.e. Websockets or gRPC-Web.

I've ended up with the following "API" for writing the chatbot:

```go
var (
		url        = os.Getenv("BOT_NEXTCLOUD_URL")
		username   = os.Getenv("BOT_NEXTCLOUD_USERNAME")
		password   = os.Getenv("BOT_NEXTCLOUD_PASSWORD")
		dbLocation = os.Getenv("BOT_DB_LOCATION")
		jitsiURL   = os.Getenv("BOT_JITSI_URL")
	)

	chatChan, statusChan := make(chan client.Chat), make(chan string)
	bot := client.NewNextcloudTalk(url, username, password, dbLocation, chatChan, statusChan)

	defer bot.Close()
	if err := bot.Open(); err != nil {
		log.Fatal(err)
	}

	go func() {
		if err := bot.ReadRooms(); err != nil {
			log.Fatal(err)
		}
	}()

	go func() {
		if err := bot.ReadChats(); err != nil {
			log.Fatal(err)
		}
	}()

	go func() {
		for status := range statusChan {
			log.Println(status)
		}
	}()

	for chat := range chatChan {
		log.Printf(`Received message from "%v" ("%v") in room "%v" with ID "%v": "%v"`, chat.ActorDisplayName, chat.ActorID, chat.Token, chat.ID, chat.Message)

		reg := regexp.MustCompile("^#video(chat|call)")

		if reg.Match([]byte(chat.Message)) {
			log.Printf(`"%v" ("%v") has requested a video call in room "%v" with ID "%v"; creating video call.`, chat.ActorDisplayName, chat.ActorID, chat.Token, chat.ID)

			bot.CreateChat(chat.Token, fmt.Sprintf("@%v started a video call. Tap on %v to join!", chat.ActorID, jitsiURL+"/"+chat.Token))
		}
	}
```

**Describe alternatives you've considered**
The `lookIntoFuture` attribute is not an option; even just "subscribing" to 15 rooms this way (and re-subscribing every 20s to get the new rooms) gets CPU usage of a 4 core, 2GB RAM cloud instance to 100%.

The framework I wrote, while it works, does polling, which is quite expensive. I've tried the attribute above, but the load is even worse with it.

**Additional context**
I did a short YouTube video (in German) for the usage of the Jitsi bot in particular; this could get you a better idea of how this could be used: https://www.youtube.com/watch?v=ahwGVSfwssM

**Comment by nickvergessen** [comment]
Yeah, currently we dont have a real chat bot API. But I guess once #1879 is there it will help a lot more.

Until then what you did as a workaround works for most cases. For more APIs check our docs for now: https://nextcloud-talk.readthedocs.io/en/latest/#nextcloud-talk-api-documentation

**Comment by nickvergessen** [comment]
> The lookIntoFuture attribute is not an option; even just "subscribing" to 15 rooms this way (and re-subscribing every 20s to get the new rooms) gets CPU usage of a 4 core, 2GB RAM cloud instance to 100%.

That should not happen, as the same happens when you have 15 users, or you mean on the bot server not the Nextcloud server?

**Comment by pojntfx** [comment]
@nickvergessen No, I mean the Nextcloud Talk server; it just goes to 100% CPU usage. Sadly I'm using a managed product (Hetzner Storageshare) so I can't open up logs ...

In case you are interested in how the API could look like, I've implemented a streaming Nextcloud Talk API here: https://github.com/pojntfx/nextcloud-talk-bot-framework/tree/add-grpc-service

It currently just proxies to a polling service, but is much simpler to use & strongly typed; here is the current gRPC proto:

```protobuf
syntax = "proto3";

package nextcloudTalk;

import "google/protobuf/empty.proto";

// NextcloudTalk is a Nextcloud Talk client.
service NextcloudTalk {
  rpc ReadChats(google.protobuf.Empty) returns (stream OutChat);
  rpc WriteChat(InChat) returns (google.protobuf.Empty);
}

message OutChat {
  int64 ID = 1;
  string Token = 2;
  string ActorID = 3;
  string ActorDisplayName = 4;
  string Message = 5;
}

message InChat {
  string Token = 1;
  string Message = 2;
}
```

This is how PubSub is used right now:

```bash
% grpcurl -plaintext 192.168.64.9:1969 nextcloudTalk.NextcloudTalk.ReadChats
{
  "ID": "50363",
  "Token": "2j9j95et",
  "ActorID": "jitsi--bot",
  "ActorDisplayName": "Jitsi Video-Chat Bot",
  "Message": "Test"
}
{
  "ID": "50366",
  "Token": "2j9j95et",
  "ActorID": "felix.pojtinger--administrator",
  "ActorDisplayName": "Felix Pojtinger (Administrator)",
  "Message": "Whoa!"
}
# ... - this is a streaming RPC
```

The proxy can also write messages easily:

```bash
% grpcurl -plaintext -d '{"Token": "2j9j95et", "Message": "Test"}' 192.168.64.9:1969 nextcloudTalk.NextcloudTalk.WriteChat
{

}
```

**Comment by pojntfx** [comment]
Just so you know, this has now evolved into a full framework with `nxtalkproxyd` (a streaming API for Nextcloud Talk) as well as a reference chatbot implementation:

- https://github.com/pojntfx/nextcloud-talk-bot-framework
- https://github.com/pojntfx/nextcloud-talk-bot-jitsi

I also did a short introduction video:

https://www.youtube.com/watch?v=WRYlHDGApZo

In case someone is looking to create a chatbot right now, this might be your best bet ;)

**Comment by DineshRefex** [comment]
hello@pojntfx help me to make a chatbot in nextcloud, please.

**Comment by nickvergessen** [comment]
Webhook based bots are available in the upcoming Nextcloud 27.1 with Talk 17.1
Preview versions of both are already available, detailed documentation can be found at https://nextcloud-talk.readthedocs.io/en/latest/bots/

Please let us know if you need any help with the API or have any questions on the usage
